odoo/models.py

```diff
@@ -2665,8 +2665,22 @@ def add(name, definition):
             if not getattr(cls, 'pool', None)
             for constraint in getattr(cls, '_local_sql_constraints', ())
         }
-
-        for (key, definition, _) in self._sql_constraints:
+        sql_constraints = list(self._sql_constraints)
+
+        if self._parent_store:
+            # we need this extension
+            cr.execute('CREATE EXTENSION IF NOT EXISTS btree_gist')
+            # Add constraint on the values of parent_left and parent_right to
+            # prevent race conditions during MPTT updates in record creation.
+            # /!\ Do not change the constraint name, as it is matched to retry
+            # transactions that failed because of them.
+            definition = 'EXCLUDE USING gist (' \
+                         'COALESCE(parent_id, 0) WITH =,' \
+                         '(CASE WHEN parent_left<parent_right THEN int4range(parent_left, parent_right+1) ELSE NULL END) WITH &&' \
+                         ') INITIALLY DEFERRED'
+            sql_constraints.append(('parent_range', definition, ""))
```

Comment:
I'll have to analyze that constraint definition further to be able to comment...
I wonder if the definition comparison system works well with this kind of definition, though? (i.e. it won't have to be re-created every time we call `auto_init`?)